modules/apps/headless/headless-discovery/headless-discovery-test/src/testIntegration/java/com/liferay/headless/discovery/jaxrs/resource/v1_0/test/HeadlessDiscoveryOpenAPIResourceTest.java

```diff
@@ -0,0 +1,184 @@
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.portal.util.PropsUtil;
+import com.liferay.portal.vulcan.util.LocalizedMapUtil;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Carlos Correa
+ */
+@RunWith(Arquillian.class)
+public class HeadlessDiscoveryOpenAPIResourceTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayIntegrationTestRule liferayIntegrationTestRule =
+		new LiferayIntegrationTestRule();
+
+	@Before
+	public void setUp() throws Exception {
+		_objectDefinition1 = _publishObjectDefinition(
+			ObjectDefinitionConstants.SCOPE_COMPANY);
+		_objectDefinition2 = _publishObjectDefinition(
+			ObjectDefinitionConstants.SCOPE_SITE);
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		_objectDefinitionLocalService.deleteObjectDefinition(
+			_objectDefinition1.getObjectDefinitionId());
+		_objectDefinitionLocalService.deleteObjectDefinition(
+			_objectDefinition2.getObjectDefinitionId());
+	}
+
+	@Test
+	public void testGetGlobalOpenAPI() throws Exception {
+		_setFeatureFlag(true, "LPS-166216");
+
+		try {
+			JSONObject globalOpenAPIJSONObject = _invoke(
+				"openapi/openapi.json");
+
+			List<String> globalOpenAPIPaths = _getPaths(
+				globalOpenAPIJSONObject);
+
+			JSONObject openAPIEndpointsJSONObject = _invoke("openapi");
+
+			Map<String, Object> openAPIEndpointsMap =
+				openAPIEndpointsJSONObject.toMap();
+
+			for (Map.Entry<String, Object> entry :
+					openAPIEndpointsMap.entrySet()) {
+
+				for (String openAPIPath : (List<String>)entry.getValue()) {
+					JSONObject openAPIJSONObject = _invoke(
+						_getOpenAPISubpath(openAPIPath));
+
+					for (String path : _getPaths(openAPIJSONObject)) {
+						if (path.endsWith("/")) {
+							path = path.substring(0, path.lastIndexOf("/"));
+						}
+
+						Assert.assertTrue(
+							globalOpenAPIPaths.remove(entry.getKey() + path));
+					}
+				}
+			}
+
+			Assert.assertTrue(
+				"The list of paths must be empty",
+				globalOpenAPIPaths.isEmpty());
+		}
+		finally {
+			_setFeatureFlag(false, "LPS-166216");
+		}
+	}
+
+	private String _getOpenAPISubpath(String openAPIPath) {
+		String openAPISubpath = StringUtil.removeFirst(
+			openAPIPath, "http://localhost:8080/o/");
+
+		return StringUtil.replaceLast(openAPISubpath, ".yaml", ".json");
+	}
+
+	private List<String> _getPaths(JSONObject openAPIJSONObject) {
+		JSONObject pathsJSONObject = openAPIJSONObject.getJSONObject("paths");
+
+		Map<String, Object> pathsMap = pathsJSONObject.toMap();
+
+		return new ArrayList<>(pathsMap.keySet());
+	}
+
+	private JSONObject _invoke(String endpoint) throws Exception {
```

Comment:
I've seen this code in many places. Would it make sense to create a common util class?

Reply:
Yeah vulcan-test-util sounds nice. And yes, you can create a separate task and add the `headless-tech-debt` label to it.

Reply:
:point_right:  https://issues.liferay.com/browse/LPS-177591